File: cpp/merge_sort.cpp
#include <iostream>
#include <vector>

void merge(std::vector<int>& arr, int const left, int const mid, int const right) {
    auto const subArrayOne = mid - left + 1;
    auto const subArrayTwo = right - mid;

    std::vector<int> leftArray(subArrayOne), rightArray(subArrayTwo);

    for (auto i = 0; i < subArrayOne; i++)
        leftArray[i] = arr[left + i];
    for (auto j = 0; j < subArrayTwo; j++)
        rightArray[j] = arr[mid + 1 + j];

    auto indexOfSubArrayOne = 0, indexOfSubArrayTwo = 0;
    int indexOfMergedArray = left;

    while (indexOfSubArrayOne < subArrayOne && indexOfSubArrayTwo < subArrayTwo) {
        if (leftArray[indexOfSubArrayOne] <= rightArray[indexOfSubArrayTwo]) {
            arr[indexOfMergedArray] = leftArray[indexOfSubArrayOne];
            indexOfSubArrayOne++;
        } else {
            arr[indexOfMergedArray] = rightArray[indexOfSubArrayTwo];
            indexOfSubArrayTwo++;
        }
        indexOfMergedArray++;
    }

    while (indexOfSubArrayOne < subArrayOne) {
        arr[indexOfMergedArray] = leftArray[indexOfSubArrayOne];
        indexOfSubArrayOne++;
        indexOfMergedArray++;
    }

    while (indexOfSubArrayTwo < subArrayTwo) {
        arr[indexOfMergedArray] = rightArray[indexOfSubArrayTwo];
        indexOfSubArrayTwo++;
        indexOfMergedArray++;
    }
}

void mergeSort(std::vector<int>& arr, int const begin, int const end) {
    if (begin >= end)
        return;

    auto mid = begin + (end - begin) / 2;
    mergeSort(arr, begin, mid);
    mergeSort(arr, mid + 1, end);
    merge(arr, begin, mid, end);
}

// Example usage
int main() {
    std::vector<int> arr = {12, 11, 13, 5, 6, 7};
    mergeSort(arr, 0, arr.size() - 1);

    for (int i : arr)
        std::cout << i << " ";
    std::cout << std::endl;

    return 0;
}

File: R/merge_sort.r
merge_sort <- function(arr) {
  if (length(arr) > 1) {
    mid <- length(arr) %/% 2
    L <- arr[1:mid]
    R <- arr[(mid+1):length(arr)]
    
    L <- merge_sort(L)
    R <- merge_sort(R)
    
    return(merge(L, R))
  } else {
    return(arr)
  }
}

merge <- function(L, R) {
  result <- c()
  while (length(L) > 0 && length(R) > 0) {
    if (L[1] <= R[1]) {
      result <- c(result, L[1])
      L <- L[-1]
    } else {
      result <- c(result, R[1])
      R <- R[-1]
    }
  }
  return(c(result, L, R))
}

# Example usage
arr <- c(12, 11, 13, 5, 6, 7)
sorted_arr <- merge_sort(arr)
print(sorted_arr)
